# Gaussian Elimination with Pivoting

In [1]:
import numpy as np

### Gaussian Elimination with Partial Pivoting

- In place of choosing the default row (row $k$) on the $k$-th pass, the  <b>pivot row</b> is chosen to have largest absolute value element ( pivot element) in the $k$-th column below and including row $k$.
 
- Row interchange takes place to make the row having the pivot element as the pivot row. (Not really in the algorithm though. A permutation vector, $p$, can keep track of the row interchanges instead. Here $p$ is initialized as $p = [1,\, 2,\, 3, \cdots,\, n]$.)
 
- The Gaussian elimination could be used for LU-decomposition  of matrix $PA$ where permutation matrix $P$ is associated with the permutation vector $p$
$$LU = PA.$$


### Gaussian elimination with scaled-row partial pivoting
---
<b>Scaled Row Partial Pivoting</b> (SRPP) is a pivoting strategy in which row-interchanges are based on the scaling of potential pivot elements by the absolute maximum element of each row.


- In Gaussian elimination with SRPP, we use a scaling vector
$s = [s_1, s_2, \cdots, s_n]$,
where
$$s_i = \max_{1  \leq j \leq n} |a_{ij}| , \qquad (1  \leq i \leq n)$$

- The pivot element (and pivot-row) at pass-$k$ is chosen based on the maximum of 
$$
\max \left \{
\frac{|a_{p_i k}|}{s_{p_i}} ;  \qquad k \leq i \leq n .
\right \}
$$
where $p$ is a permuations vectors that keeps track of row interchanges. It is initialized with default ordering $[1,\ 2,\ 3, \cdots \ n ]$. Rows swaps are logically performed by intercahnging the elements pf $p$. For example if row 1 is changed with row 2, then $p$ changes to $p=[2,\ 1,\ 3, \ \cdots \ n ]$.
 
- The Gaussian elimination could again be used for LU-decomposition  as 
$$LU = PA,$$
where is $P$ is the permutation matrix associated with vector $p$, given by 
$$
P=I_n[p,\,:]
$$
For example, if $p=[3,1,2]$, the corresponding permutation matrix is given by

$$
P=I_3[p,\,:] =I_3[[3,1,2],\,:] =  \begin{bmatrix}
0 & 0 & 1  \\
1 & 0 & 0  \\
0 & 1 & 0 
\end{bmatrix}
$$

The following code provides an implementation of Gaussian Elimination with scaled row-partial pivoting. Ponder over the following.

  - What is the need for pivoting?
  - How can you find a factorization of the matrix by using the following code.
  - How can you modify the code to find the determinant of a matrix?
  - How can you modify the following code to solve a linear system?
  - How can you modify the code to find the inverse of a matrix?

In [3]:
## Gaussian Elimination: with Scaled Row Pivoting
## This function is based on the pseudo-code on page-148 in the Text by Kincaid and Cheney
def GE_srpp(X, verbose=False):
    '''
    This function returns the P'LU factorization of a square matrix A
    by scaled row partial pivoting. 
    In place of returning L and U, elements of modified A are used to hold values of L and U.
    '''
    A = np.copy(X)
    m,n = A.shape
    swap=0;
    
    # The initial ordering of rows
    p = list(range(m))
    if verbose:
        print("permutation vector initialized to: ",p)
    
    # Scaling vector: absolute maximum elements of each row
    s = np.max(np.abs(A), axis=1) 
    
    # Start the k-1 passes of Guassian Elimination on A
    for k in range(m-1):              
        if verbose:
            print("Scaling Vector: ",s)
        # Find the pivot element and interchange the rows
        pivot_index = k + np.argmax(np.abs(A[p[k:], k])/s[p[k:]])        
        
        # Interchange elements in the permutation vector if needed
        if pivot_index !=k:
            temp = p[k]
            p[k]=p[pivot_index]
            p[pivot_index] = temp
            swap+=1;
                        
        if verbose:
            print("\nPivot Element: {0:.4f} \n".format(A[p[k],k]))
        if np.abs(A[p[k],k]) < 10**(-20):
             sys.exit("ERROR!! Provided matrix is singular or there is a zero in pivot position.")        
        # Check the new order of rows
        if verbose:
            print("permutation vector: ",p)
        # For the k-th pivot row Perform the Gaussian elimination on the following rows
        for i in range(k+1, m):
            # Find the multiplier
            z = A[p[i],k]/A[p[k],k]            
            #Save the multiplier z in A itself. You can save this in L also
            A[p[i],k] = z          
            #Elimination operation: Changes all elements in a row simultaneously
            A[p[i],k+1:] = A[p[i],k+1:] - z*A[p[k],k+1:]
            
        if verbose:
            #print("\n After PASS {}=========: \n".format(k+1), A)
            print("\n After PASS {}=========: \n".format(k+1), bmatrix(A))
            
    return A, p, swap

In [10]:
# In the following function A is supposed to be upper triangular systems UX=b
def back_sub(U, b):
    n = U.shape[0]
    #Solution will be saved in variable x
    x = np.zeros_like(b, dtype=float)

    # Back-substitution
    ##last variable is found first
    x[n-1] = b[n-1] / U[n-1,n-1]
    ## Find the remaining n-1 variables from last to first
    for k in range(n-2,-1,-1):
        known_sums = np.dot(U[k,k+1:],x[k+1:])
        x[k] = (b[k] - known_sums) / U[k,k]
    return x

In [4]:
## Example on page number 146 (Kincaid Cheney).
## Example solved manually in class
A0 = np.array([[2, 3, -6], [1,-6,8], [3, -2, 1]], dtype=float)
#A = np.array([[5, 4, 7, 6, 9], [7, 8, 9, 9, 8], [2, 3, 5, 9, 8], [3, 1, 7, 5, 6], [9, 1, 3, 7, 3]], dtype=float)
#A0=np.array([[1, 0, 2, 1],[4, -9, 2, 1],[8, 16, 6, 5],[2, 3, 2,1]], dtype=float)
print("\n Given A: \n ",bmatrix(A0))


 Given A: 
  \begin{bmatrix}
  2. & 3. & -6.\\
  1. & -6. & 8.\\
  3. & -2. & 1.\\
\end{bmatrix}


In [5]:
A,p,swap =GE_srpp(A0,verbose=True) #You need to pass a copy, otherwise the original A0 is changed.
m,n=A.shape

permutation vector initialized to:  [0, 1, 2]
Scaling Vector:  [6. 8. 3.]

Pivot Element: 3.0000 

permutation vector:  [2, 1, 0]

 After PASS 1=========: 
 \begin{bmatrix}
  0.66666667 & 4.33333333 & -6.66666667\\
  0.33333333 & -5.33333333 & 7.66666667\\
  3. & -2. & 1.\\
\end{bmatrix}
Scaling Vector:  [6. 8. 3.]

Pivot Element: 4.3333 

permutation vector:  [2, 0, 1]

 After PASS 2=========: 
 \begin{bmatrix}
  0.66666667 & 4.33333333 & -6.66666667\\
  0.33333333 & -1.23076923 & -0.53846154\\
  3. & -2. & 1.\\
\end{bmatrix}


Here is a visualization for the above Gaussian elimination process with scaled row partial pivoting.

$$
 \begin{bmatrix}
  2. & 3. & -6.\\
  1. & -6. & 8.\\
  \color{red}{[3]}. & -2. & 1.\\
\end{bmatrix} \longrightarrow 
\begin{bmatrix}
  \color{blue}{(0.667)} & \color{red}{[4.333]} & -6.667\\
  \color{blue}{(0.333)} & -5.333 & 7.667\\
  \color{red}{[3]}. & -2. & 1.\\
\end{bmatrix}\longrightarrow 
\begin{bmatrix}
  \color{blue}{(0.667)} & \color{red}{[4.333]} & -6.667\\
  \color{blue}{(0.333)} & \color{blue}{(-1.231)} & -0.538\\
  \color{red}{[3]}. & -2. & 1.\\
\end{bmatrix}
$$

Note: The numbers of the form $\color{red}{[3]}$ are pivot elements.
The numbers of the form $\color{blue}{(0.667)}$ are zeros that are now storing the multipliers from the Gaussian Elimination.


In [7]:
L = np.tril(A[p,:], -1)+np.eye(n)
U = np.triu(A[p,:]) # YOu can solve Ux = b[p] by back substitution
P = np.eye(n)[p,:]
print("\n After Gaussian Elimination with SRPP: \n", A)
print("\n The permutation Vector is: \n", p)
print("\n Permutation matrix, P:\n ", P)
print("\n Upper triangular, U:\n ", U)
print("\n Lower triangular, L:\n", L)
print("Sanity Check: norm of PA-LU",np.linalg.norm(P@A0-L@U))


 After Gaussian Elimination with SRPP: 
 [[ 0.66666667  4.33333333 -6.66666667]
 [ 0.33333333 -1.23076923 -0.53846154]
 [ 3.         -2.          1.        ]]

 The permutation Vector is: 
 [2, 0, 1]

 Permutation matrix, P:
  [[0. 0. 1.]
 [1. 0. 0.]
 [0. 1. 0.]]

 Upper triangular, U:
  [[ 3.         -2.          1.        ]
 [ 0.          4.33333333 -6.66666667]
 [ 0.          0.         -0.53846154]]

 Lower triangular, L:
 [[ 1.          0.          0.        ]
 [ 0.66666667  1.          0.        ]
 [ 0.33333333 -1.23076923  1.        ]]
Sanity Check: norm of PA-LU 0.0


In [8]:
A0
print(A0)
b=np.array([-1,3,2], dtype=float).reshape(3,-1)
augA = np.hstack((A0,b))
print(augA)
print(np.hstack((A0,np.eye(3))))

[[ 2.  3. -6.]
 [ 1. -6.  8.]
 [ 3. -2.  1.]]
[[ 2.  3. -6. -1.]
 [ 1. -6.  8.  3.]
 [ 3. -2.  1.  2.]]
[[ 2.  3. -6.  1.  0.  0.]
 [ 1. -6.  8.  0.  1.  0.]
 [ 3. -2.  1.  0.  0.  1.]]


In [11]:
newA,p,swap = GE_srpp(augA)
newU = np.triu(newA[p,:3])
newb = newA[p,3]
print(back_sub(newU, newb))

[1. 1. 1.]


In [12]:
A_test = np.array([[3, -5, 1], 
              [1, 2, 3], 
              [-2, 3, -4]], dtype=float)
b = np.array([[0],[1],[3]], dtype=float)
augA = np.hstack((A_test,b))
newAug,p,swaps = GE_srpp(augA,verbose=True)

permutation vector initialized to:  [0, 1, 2]
Scaling Vector:  [5. 3. 4.]

Pivot Element: 3.0000 

permutation vector:  [0, 1, 2]

 After PASS 1=========: 
 \begin{bmatrix}
  3. & -5. & 1. & 0.\\
  0.33333333 & 3.66666667 & 2.66666667 & 1.\\
  -0.66666667 & -0.33333333 & -3.33333333 & 3.\\
\end{bmatrix}
Scaling Vector:  [5. 3. 4.]

Pivot Element: 3.6667 

permutation vector:  [0, 1, 2]

 After PASS 2=========: 
 \begin{bmatrix}
  3. & -5. & 1. & 0.\\
  0.33333333 & 3.66666667 & 2.66666667 & 1.\\
  -0.66666667 & -0.09090909 & -3.09090909 & 3.09090909\\
\end{bmatrix}


In [ ]:
print("The Permutation Matrix, P:\n",P)
print("Error Norm \n", np.linalg.norm(np.dot(L,U) - np.dot(P,A0)))
print(np.dot(L,U))
print(np.dot(P,A0))

### P LU Factorization 
---
One can use the Gaussian elimination with partial pivoting to decompose the matrix A as follows
$$
P A = L U,\text{ or }\ A = P^T L U;
$$
where $P$ is a permutation matrix, $L$ is a unit lower-triangular matrix and $U$ is an upper triangular matrix.

## An Example From LLM
Solving a linear system using Gaussian elimination with scaled row partial pivoting involves several steps. The goal is to transform the system into an upper triangular form, which can then be solved using back substitution. Scaled row partial pivoting helps to reduce numerical errors by selecting the pivot element with the largest relative magnitude in the column.

### Example System:
Consider the following system of linear equations. <strong>There are som errors in explanation of the algorithm. Find them out and correct them.</strong> 

$$
\begin{cases}
2x + 3y - z = 5 \\
4x + 4y - 3z = 3 \\
2x - 3y + z = -1
\end{cases}
$$

This can be written in matrix form as:

$$
\begin{bmatrix}
2 & 3 & -1 & | & 5 \\
4 & 4 & -3 & | & 3 \\
2 & -3 & 1 & | & -1
\end{bmatrix}
$$

### Step 1: Compute the Scaling Factors
For each row, compute the scaling factor $ s_i$, which is the maximum absolute value in that row (excluding the augmented column).

$$
s_1 = \max{(|2|, |3|, |-1|)} = 3 \\
s_2 = \max{(|4|, |4|, |-3|)} = 4 \\
s_3 = \max{(|2|, |-3|, |1|)} = 3
$$

### Step 2: Select the Pivot Row
For the first column, compute the ratios $|a_{i1}|/s_i$ for each row $ i $:

$$
\text{Ratio for Row 1: } |2| / 3 = 0.6667 \\
\text{Ratio for Row 2: } |4| / 4 = 1.0 \\
\text{Ratio for Row 3: } |2| / 3 = 0.6667
$$

The largest ratio is 1.0, corresponding to Row 2. Therefore, Row 2 is the pivot row for the first step.

### Step 3: Swap Rows (if necessary)
Since Row 2 is already in the second position, no swapping is needed.

### Step 4: Eliminate the First Column Below the Pivot
Use the pivot element \( a_{21} = 4 \) to eliminate the first column elements below it.

- **Row 1:** \( R_1 = R_1 - (a_{11}/a_{21}) \times R_2 = R_1 - (2/4) \times R_2 \)
- **Row 3:** \( R_3 = R_3 - (a_{31}/a_{21}) \times R_2 = R_3 - (2/4) \times R_2 \)

Calculating these:

$$
R_1 = R_1 - 0.5 \times R_2 = [2, 3, -1, 5] - 0.5 \times [4, 4, -3, 3] = [0, 1, 0.5, 3.5] \\
R_3 = R_3 - 0.5 \times R_2 = [2, -3, 1, -1] - 0.5 \times [4, 4, -3, 3] = [0, -5, 2.5, -2.5]
$$

The updated matrix is:

$$
\begin{bmatrix}
0 & 1 & 0.5 & | & 3.5 \\
4 & 4 & -3 & | & 3 \\
0 & -5 & 2.5 & | & -2.5
\end{bmatrix}
$$

### Step 5: Select the Pivot Row for the Second Column
Now, consider the second column below the first pivot row. Compute the ratios $ |a_{i2}| / s_i $ for each remaining row $ i $:

$$
\text{Ratio for Row 1: } |1| / 3 = 0.3333 \\
\text{Ratio for Row 3: } |-5| / 3 = 1.6667
$$

The largest ratio is 1.6667, corresponding to Row 3. Therefore, Row 3 is the pivot row for the second step.

### Step 6: Swap Rows (if necessary)
Swap Row 1 and Row 3:

$$
\begin{bmatrix}
0 & -5 & 2.5 & | & -2.5 \\
4 & 4 & -3 & | & 3 \\
0 & 1 & 0.5 & | & 3.5
\end{bmatrix}
$$

### Step 7: Eliminate the Second Column Below the Pivot
Use the pivot element $ a_{32} = 1$ to eliminate the second column elements below it.

- **Row 1:** $ R_1 = R_1 - (a_{12}/a_{32}) \times R_3 = R_1 - (-5/1) \times R_3 $

Calculating this:

$$
R_1 = R_1 + 5 \times R_3 = [0, -5, 2.5, -2.5] + 5 \times [0, 1, 0.5, 3.5] = [0, 0, 5, 15]
$$

The updated matrix is:

$$
\begin{bmatrix}
0 & 0 & 5 & | & 15 \\
4 & 4 & -3 & | & 3 \\
0 & 1 & 0.5 & | & 3.5
\end{bmatrix}
$$

### Step 8: Back Substitution
Now, the matrix is in upper triangular form. Solve for the variables starting from the bottom:

1. From Row 3: \( 0x + 1y + 0.5z = 3.5 \) → \( y + 0.5z = 3.5 \)
2. From Row 2: \( 4x + 4y - 3z = 3 \)
3. From Row 1: \( 0x + 0y + 5z = 15 \) → \( z = 3 \)

Substitute $ z = 3$ into Row 3:

$$
y + 0.5(3) = 3.5 \rightarrow y = 2
$$

Substitute $ y = 2 $ and $ z = 3 $ into Row 2:

$$
4x + 4(2) - 3(3) = 3 \rightarrow 4x + 8 - 9 = 3 \rightarrow 4x = 4 \rightarrow x = 1
$$

### Final Solution:
$$
x = 1, \quad y = 2, \quad z = 3
$$

This is the solution to the system of linear equations using Gaussian elimination with scaled row partial pivoting.

In [2]:
def bmatrix(A):
    """This function converts a 1-2 dimensional array into a LaTeX B-matrix
    INPUT:a: numpy array
    OUTPUT: LaTeX B-matrix code
    """
    if len(A.shape) > 2:
        raise ValueError('The input array must be a vector or a matrix.')
    
    rows = str(A).replace('[', '').replace(']', '').splitlines()
    ltx_str = [r'\begin{bmatrix}']
    ltx_str += ['  ' + ' & '.join(val.split()) + r'\\' for val in rows]
    ltx_str +=  [r'\end{bmatrix}']
    return '\n'.join(ltx_str)

A = np.array([[12, 5, 2], [20, 4, 8], [ 2, 4, 3], [ 7, 1, 10]])
print (bmatrix(A) + '\n')

B = np.array([[1.2], [3.7], [0.2]])
print (bmatrix(B) + '\n')

C = np.array([1.2, 9.3, 0.6, -2.1])
print (bmatrix(C) + '\n')

\begin{bmatrix}
  12 & 5 & 2\\
  20 & 4 & 8\\
  2 & 4 & 3\\
  7 & 1 & 10\\
\end{bmatrix}

\begin{bmatrix}
  1.2\\
  3.7\\
  0.2\\
\end{bmatrix}

\begin{bmatrix}
  1.2 & 9.3 & 0.6 & -2.1\\
\end{bmatrix}

